misk/src/main/kotlin/misk/client/HttpClientsConfig.kt

```diff
@@ -1,115 +1,75 @@
-  val defaultPingInterval: Duration? = null,
-  val defaultCallTimeout: Duration? = null,
-  @JsonDeserialize(contentConverter = BackwardsCompatibleEndpointConfigConverter::class)
+  @JsonAlias("hosts")
+  //Need to retain ordering, hence LinkedHashMap
+  val hostConfigs: LinkedHashMap<String, HttpClientConfig> = linkedMapOf(),
   val endpoints: Map<String, HttpClientEndpointConfig> = mapOf()
 ) : Config {
+  init {
+    validatePatterns()
+  }
+
+  private fun validatePatterns() = try {
+    endpoints.keys
+        .map { it.toRegex(RegexOption.IGNORE_CASE) }
+        .forEach { it.matches("") }
+  } catch (e: Exception) {
+    throw IllegalArgumentException(
+        "Failed to initialize HttpClientsConfig, failed to parse Regexp patterns!",
+        e
+    )
+  }
+
   /** @return The [HttpClientEndpointConfig] for the given client, populated with defaults as needed */
   operator fun get(clientName: String): HttpClientEndpointConfig {
-    // TODO(mmihic): Cache, proxy, etc
-    val endpointConfig = endpoints[clientName] ?: throw IllegalArgumentException(
-        "no client configuration for endpoint $clientName")
+    val endpointConfig = requireNotNull(endpoints[clientName]) {
+        "no client configuration for endpoint $clientName"
+    }
+
+    val allMatchingConfigs = sequence {
+      yield(httpClientConfigDefaults)
+      yieldAll(
+          endpointConfig.url?.let { url ->
+            findUrlMatchingConfigs(url)
+          } ?: findWildcardConfigs()
+      )
+      yield(endpointConfig.clientConfig)
+    }.reduce { prev, cur -> cur.applyDefaults(prev) }
 
-    val clientConfig = endpointConfig.clientConfig
     return HttpClientEndpointConfig(
         url = endpointConfig.url,
         envoy = endpointConfig.envoy,
-        clientConfig = clientConfig.copy(
-            connectTimeout = clientConfig.connectTimeout ?: defaultConnectTimeout,
-            writeTimeout = clientConfig.writeTimeout ?: defaultWriteTimeout,
-            readTimeout = clientConfig.readTimeout ?: defaultReadTimeout,
-            pingInterval = clientConfig.pingInterval ?: defaultPingInterval,
-            callTimeout = clientConfig.callTimeout ?: defaultCallTimeout,
-            ssl = clientConfig.ssl ?: ssl
-        )
+        clientConfig = allMatchingConfigs
     )
   }
 
-  /**
-   * Returns a default config with no url/envoy config specified.
-   * Used to build up endpoint config for clients dynamically.
-   */
-  fun getDefault(): HttpClientEndpointConfig {
-    return HttpClientEndpointConfig(
-        url = null,
-        envoy = null,
-        clientConfig = HttpClientConfig(
-            connectTimeout = defaultConnectTimeout,
-            writeTimeout = defaultWriteTimeout,
-            readTimeout = defaultReadTimeout,
-            pingInterval = defaultPingInterval,
-            callTimeout = defaultCallTimeout,
-            ssl = ssl
-        )
+  private fun findWildcardConfigs() =
+      hostConfigs.filter { (k, _) -> k == ".*" }.values
```

Comment:
This is strictly for backwards compatibility to support `envoy` endpoints, for which we do not have URL to match against.